### PR TITLE
Use SRWLOCK in atomic_load operation on shared_ptr

### LINKED ISSUE #370: atomic.cpp: Spinlock powering shared_ptr atomics can lead to priority inversion
**Describe the bug**
shared_ptr's atomic functions (e.g. `std::atomic_store`) are powered by an external lock present in our separately compiled machinery, `_Lock_shared_ptr_spin_lock` and `_Unlock_shared_ptr_spin_lock`. This was written to use a plain spin lock with no mitigation to go to sleep if spinning is taking too long, nor is there any mitigation for memory bandwidth consumption.

https://github.com/microsoft/STL/blob/aa0a7a3d859ade0f6f1ff13aa4ef74b3d5ce2326/stl/src/atomic.cpp#L13-L34

In single threaded scenarios this is particularly bad when a low priority thread currently holds the spinlock, and a high priority thread spins "effectively forever".

In an ABI breaking release, it would be nice to reuse the low order bit of the reference count control block pointer in the shared_ptr; but even without an ABI breaking release we could do better by replacing the spinlock entirely with something like `SRWLOCK` on Vista and later, implementing exponential backoff, or relying on C++20 `std::atomic` waiting features once we have those implemented.

Also tracked by Developer Community as DevCom-716238 and Microsoft-internal VSO-975564.

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Fix #370

### CODE REVIEW

File: stl/src/atomic.cpp [hunk 1]
```diff
@@ -5,32 +5,21 @@
...
+namespace {
+    // MUTEX FOR shared_ptr ATOMIC OPERATIONS
+    SRWLOCK _Shared_ptr_lock = SRWLOCK_INIT;
+} // unnamed namespace
 
-// SPIN LOCK FOR shared_ptr ATOMIC OPERATIONS
-volatile long _Shared_ptr_flag;
+_EXTERN_C
 
 _CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // spin until _Shared_ptr_flag successfully set
```
Comment (StephanTLavavej) [inline stl/src/atomic.cpp:17]: This comment refers to a variable that no longer exists. I recommend replacing it with a comment about the name of the function:
```suggestion
_CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // TRANSITION, ABI: "spin_lock" name is outdated
```

**Comment by StephanTLavavej** [comment]
Thanks for improving performance here and fixing the priority inversion bug! :rocket:

**Comment by sonyps5201314** [comment]
This PR makes CRT no longer compatible with XP, but 14.27 and earlier versions are fine

**Comment by miscco** [comment]
> This PR makes CRT no longer compatible with XP, but 14.27 and earlier versions are fine

As far as I understand the whole support for XP was ripped out recently so you should not expect recent versions of the STL to work on XP

**Comment by sonyps5201314** [comment]
Why not remove XP running support when a new VS major release is released, instead choosing a minor version of VS2019 that is also undeclared and confusing to users?

**Comment by sonyps5201314** [comment]
> > This PR makes CRT no longer compatible with XP, but 14.27 and earlier versions are fine
> 
> As far as I understand the whole support for XP was ripped out recently so you should not expect recent versions of the STL to work on XP

Yesterday’s VS2019 16.8.0, VC 14.28.29333 did not quote the STL code from September 26 and later, so it should be able to maintain compatibility with the XP system.

**Comment by BillyONeal** [comment]
@sonyps5201314 We can no longer support XP, because XP doesn't support SHA2 code signing certificates, the SHA1 code signing certificates have expired, and the folks who own crypto refuse to mint new SHA1 code signing certificates. We're going to have to leave the last redist that supported XP available but you should not expect future STLs to work there.

Note that the VS2019 build tools officially dropped support for targeting XP following their deprecation in VS2017. (The IDE still supports targeting XP, but by using the VS2017 build tools options)

**Comment by sonyps5201314** [comment]
@BillyONeal 
The name "Microsoft Visual C++ 2015-2019 Redistributable" will make users think that this installation package can be safely upgraded in place to replace the previously released "Microsoft Visual C++ 2015 Redistributable" and "Microsoft Visual C++ 2017 Redistributable", since 14.27 and earlier CRT versions is still compatible with XP,
Why not consider removing the CRT's support for XP in the next new major version of VS?

**Comment by BillyONeal** [comment]
@sonyps5201314 We are no longer capable of shipping software that can run on XP, because the signing certificates expired. The signing certificates expiring was not a thoughtfully considered decision, it was something happened because signing certificates have life spans. There's nothing the Visual C++ team can do about that without some form of time machine to go back to 2002 and give XP SHA2 certificate support (and to deploy unexpired root certificates to the world's XP machines).

>since 14.27 and earlier CRT versions is still compatible with XP,

All support in the product to target XP was removed in 2019 RTM. (there is a v141_xp PlatformToolset but not a v142_xp one) Some parts of the product, including the Visual C++ team's redist DLLs, were updates to corresponding components from previous versions of the product which did support XP. As a result, as long as we could, we maintained backwards compatibility to XP there to simplify deployment for those older versions of the product.

**Comment by sonyps5201314** [comment]
@BillyONeal
Well, in fact, whether the file contains a digital signature is not very important. The early versions of CRT provided by Microsoft, such as VS2003 and VS2005 CRT, do not have a digital signature, and users can judge whether the CRT file is unmodified by the HASH value such as MD5.
Thanks for your attention.

**Comment by BillyONeal** [comment]
>Well, in fact, whether the file contains a digital signature is not very important.

I don't believe we can legally ship unsigned code. But this is getting close to legal advice so I'm not going into specifics.

>The early versions of CRT provided by Microsoft, such as VS2003 and VS2005 CRT, do not have a digital signature

That's not true. They are catalog signed, but they are still signed.

>UCRT, ATL and MFC are currently XP compatible，even the MFC program written by VS2019 can be easily compatible with XP

They are in "happens to work" territory, not "supported" territory.

**Comment by sonyps5201314** [comment]
OK, thank you for your patient explanation.

**Pull request merged**